pkg/identity/nonglobal/temp_id.go

```diff
@@ -0,0 +1,178 @@
+		idToIdentity:      make(map[identity.NumericIdentity]*identity.Identity),
+		keyToIdentity:     make(map[string]*identity.Identity),
+		markedForDeletion: make(map[identity.NumericIdentity]bool),
+	}
+}
+
+// FindOrCreateTempID gets ans existing or creates a temporary security identity
+// for the specified labels, to be used before operator creates a global
+// security identity.
+func (a *TempSecIDAllocator) FindOrCreateTempID(lbls labels.Labels) (*identity.Identity, error) {
+	if id, exists := a.LookupByIDKey(&key.GlobalIdentity{LabelArray: lbls.LabelArray()}); exists {
+		return id, nil
+	}
+
+	allocatedID, err := a.allocator.AllocateRandom()
+	if err != nil {
+		return nil, err
+	}
+
+	numID := identity.NumericIdentity(int64(allocatedID))
+	id := identity.NewIdentity(numID, lbls)
+	a.insertToCache(id)
+
+	return id, nil
+}
+
+func (a *TempSecIDAllocator) LookupByID(numID identity.NumericIdentity) (*identity.Identity, bool) {
+	if !a.allocator.IsInPoolRange(idpool.ID(numID)) {
+		return nil, false
+	}
+
+	cache := a.tempIDCache
+	cache.mu.RLock()
+	defer cache.mu.RUnlock()
+
+	id, exists := cache.idToIdentity[numID]
+	return id, exists
+}
+
+func (a *TempSecIDAllocator) LookupByIDKey(idKey *key.GlobalIdentity) (*identity.Identity, bool) {
+	cache := a.tempIDCache
+	cache.mu.RLock()
+	defer cache.mu.RUnlock()
+
+	id, exists := cache.keyToIdentity[idKey.GetKey()]
+	return id, exists
+}
+
+func (a *TempSecIDAllocator) insertToCache(id *identity.Identity) {
+	numID := id.ID
+	if !a.allocator.IsInPoolRange(idpool.ID(numID)) {
+		return
+	}
+
+	cache := a.tempIDCache
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	idKey := &key.GlobalIdentity{LabelArray: id.LabelArray}
+	cache.idToIdentity[numID] = id
+	cache.keyToIdentity[idKey.GetKey()] = id
+}
+
+func (a *TempSecIDAllocator) StartPeriodicGC(stopChan chan struct{}) {
+	log.Info("Starting Temp ID periodic garbage collection")
+
+	wait.Until(a.runGC, TempIDGCInterval, stopChan)
+
+	log.Info("Stopping Temp ID periodic garbage collection")
+}
+
+func (a *TempSecIDAllocator) runGC() {
+	gcStartTime := time.Now()
+	log.Debug("Running Temp ID garbage collection")
+	deletedCount := 0
+	defer func() {
+		gcDuration := time.Since(gcStartTime)
+		log.Debugf("Completed Temp ID periodic garbage collection after %v. Deleted Temp IDs: %d", gcDuration, deletedCount)
+	}()
+
+	eps := a.endpointListerFunc()
+	if len(eps) == 0 {
+		return
+	}
+
+	usedTempIDs := make(map[identity.NumericIdentity]bool)
+	for _, ep := range eps {
+		if ep == nil || ep.SecurityIdentity == nil {
+			continue
+		}
+
+		if identity.IsTempID(ep.SecurityIdentity.ID) {
+			usedTempIDs[ep.SecurityIdentity.ID] = true
+		}
+	}
+
+	cache := a.tempIDCache
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	for numID, id := range cache.idToIdentity {
+		if _, isUsed := usedTempIDs[numID]; isUsed {
+			delete(cache.markedForDeletion, numID)
+			continue
+		}
+
+		if markedforDeletion := cache.markedForDeletion[numID]; !markedforDeletion {
```

Comment:
I think it's possible that ID is leaked, marked for deletion and then after you list endpoints but before you lock a cache it may be reused.

When id is reused it needs to clear mark for deletion.

Reply:
I moved the lock to the start of the GC func, and also I had to move the temp ID cache lock, from insertToCache to FindOrCreateTempID, to ensure that there won't be leaks.